### ISSUE TITLE
Add text decorations

### ISSUE THREAD

**Issue by gyscos** [gyscos/Cursive#90]
Allow a TextView to have special effects on parts of the content only: bold/italic/underline, different color/highlight...

**Comment by dermetfan** [comment]
I [started](https://github.com/dermetfan/Cursive/commit/026eb9c28cfb5b2463e041e5b773323d2bd878bf) working on this. Maybe you can have a look so we agree on the direction to take this in?

You can test the current state by dropping the following in line 244 of `TextView` and running the `lorem` example. It just splits the text at "sit" and applies some style. (Yes, this is horrible and only works with the lorem ipsum text)
```rust
let delim = "sit";
let mut reappend_delim_i = 0;
let parts = text.split(delim)
    .map(|it| {
        let ret = if reappend_delim_i != 0 {
            format!("{}{}", delim, it)
        } else { it.to_owned() };
        reappend_delim_i += 1;
        ret
    })
    .collect::<Vec<_>>();
let parts = parts.iter()
    .map(|it| it.as_str())
    .collect::<Vec<_>>();
if parts.len() == 3 {
    let part1 = (parts[0], ::style::Style {
        color: None,
        ..Default::default()
    });
    let part2 = (parts[1], ::style::Style {
        color: Some(::theme::ColorStyle::Highlight),
        ..Default::default()
    });
    let part3 = (parts[2], ::style::Style {
        effect: ::theme::Effect::Reverse,
        ..Default::default()
    });
    printer.print_styled((x, 0), &[part1, part2, part3][..]);
} else {
    printer.print((x, 0), text);
}
```

**Comment by gyscos** [comment]
I'd like, if possible, to separate the parsing part (where marked up text is parsed into a style tree) and the rendering part, to allow the same `TextView` to be easily filled from HTML, markdown or some other text format.

**Comment by dermetfan** [comment]
I was planning to do parsing in its own module where we could implement different parsers for different markup languages.

I don't think we need a tree to store a parsing result. That is only useful if you want to modify the tree (like JavaScript inserting DOM nodes) but that's probably overkill so we can just parse the new markup string instead. The result could be stored as `Vec<StyledStr<'a>>` as you said in https://github.com/gyscos/Cursive/pull/106#commitcomment-20684775.

My current draft is

```rust
/// Handle parsing markup into style.
pub mod markup {
    /// Parses a markup string into `StyledStr`s.
    pub type Parse = for<'a> FnOnce(&'a str) -> Vec<StyledStr<'a>>;

    /// Cursive's own markup language.
    pub fn parse_cursive<'a>(markup: &'a str) -> Vec<StyledStr<'a>> {
        unimplemented!()
    }
}
```

which would be used in widgets like

```rust
// If `None`, disable markup.
markup: Option<Box<markup::Parse>>,
```

> I'm still trying to find how to design it around use-cases like TextView where styled spans are rows are overlapping...

After pondering quite a while I realized that the `LinesIterator` that computes `rows` must obviously not operate on the markup string but on the parsed result. That means we need to adapt computation and usage of `rows` anyway.

I haven't yet thought about this in depth but my first idea is to write a `StyledLinesIterator` or so with a `type Row<'a> = &'a [StyledStr<'a>]`. Maybe that `StyledLinesIterator` could use the already existing `LinesIterator` so it doesn't need to be completely rewritten.

I'm out earning money until Friday. Looking forward to continue with this.

**Comment by gyscos** [comment]
Let's start from the end.
* The `Printer` takes a `&[StyledStr<'a>]` per row.
* We want to store the rows (to avoir re-computing them) but storing these `StyledStr<'a>` inside `TextView` might be tricky (self-borrowing types are never easy to deal with).
So instead, we'll store a list of `struct StyledSpan { style: Style, start: usize, end: usize, width: usize }`. It's then trivial to build the list of `StyledStr<'a>` from that when drawing.
* Our `StyledLinesIterator` should then output `StyledSpan`s. It will need to:
  * Accumulate `StyledSpan`s as long as they fit on one row
  * Split the last span following the current rules
  * Start the next row with the leftovers from the split span.
* The input to the `StyledLinesIterator` would then be a styled input, made of:
  * A list of `StyledSpan`s
  * The input string the spans are indexing into

  (Keeping the string separate and using indices in there is just a trick to avoid issues with storing references in the string; if there is another option - I haven't checked [Kimundi/owning-ref-rs](https://github.com/Kimundi/owning-ref-rs) yet - that could change.)

So all in all, a `TextView` would have to store:
* The main input string
* The result of the style parsing: a list of `StyledSpan`
* Pre-computed rows: a list of lists of `StyledSpan`

And when asked to draw itself, it would map a list of `StyledSpan` to a list of `StyledStr<'a>` using the input string, and give that to the printer...

**Comment by dermetfan** [comment]
That sounds good! I'll start working on it when I have some time.

I think our `parse()` signature from [dermetfan@`bf43333` (comment)](https://github.com/dermetfan/Cursive/commit/bf43333a3b25324d8603ae68b9458820f6fb19c7#commitcomment-20684775) is unable to parse markup with escape sequences.

Consider this HTML: `<i>1 &lt; 2</i>`

Using this signature, our parser could only return:
```rust
[(
  "1 &lt; 2",
  Style {
    color: None,
    effect: Effect::Italic
  }
)]
```

Since `parse()` just has a string slice it cannot transform this into:
```rust
[(
  "1 < 2",
   ...
)]
```

A `Cow` solves the problem: `parse<'a>(markup: &'a mut Cow<str>) -> Vec<StyledStr<'a>>`

So to update the example from my [previous comment](https://github.com/gyscos/Cursive/issues/90#issuecomment-276542894):
```rust
/// Handle parsing markup into style.
pub mod markup {
    use super::*;
    use std::borrow::Cow;

    /// Parses a markup string into `StyledStr`s.
    pub type Parse = for<'a> FnOnce(&'a mut Cow<str>) -> Vec<StyledStr<'a>>;

    /// Cursive's own markup language.
    pub fn parse_cursive<'a>(markup: &'a mut Cow<str>) -> Vec<StyledStr<'a>> {
        unimplemented!()
    }
}
```

**Comment by gyscos** [comment]
Aah damn, I forgot about that. `Cow` seems perfectly indicated in this case indeed.

**Comment by gyscos** [comment]
Oh wait - should the input itself be `Cow`, or just the content of `StyledStr<'a>`?

**Comment by dermetfan** [comment]
I meant the input into `parse()` but yes `type StyledStr<'a> = (Cow<'a, str>, Style)` is probably better. That would allow to easily parse the tags producing a `[StyledStr]` and hand that off to a function that replaces the escape sequences. We get an owned version for free :)

**Comment by spacekookie** [comment]
Any update on this feature?

**Comment by gyscos** [comment]
Yes, I've been working on the line-breaking algorithm for the past few days, and I almost got a working version. Should be able to push something this week.

**Comment by gyscos** [comment]
There is now `utils::lines::span::SpanLinesIterator`, which takes as input a list of spans (a `Cow<str>` + `Style`) and outputs rows, where each row is a list of segment (a span ID + start/end byte offsets). Given the original list of spans, a row can be resolved to a list of spans, which is easy to print.

The next steps will be:
* Add a module to parse marked-up text into spans
* Update `TextView` to work with spans instead of a single `String` (will need owning-ref here to store the `Span<'a>` next to the input `String`).
* Cross fingers

**Comment by gyscos** [comment]
I got a prototype working: you can check [examples/markup.rs](https://github.com/gyscos/Cursive/blob/master/examples/markup.rs), there's a basic markdown usage.

![screenshot from 2018-01-10 23-47-33](https://user-images.githubusercontent.com/564903/34800712-576be18e-f665-11e7-9516-e1ef462302b9.png)

Notes:
* It requires the `markdown` feature to enable the markdown parser
* The ncurses backend currently lacks italic (see https://github.com/jeaye/ncurses-rs/issues/144), but pancurses and termion should be fine.
* We now rely on a fork of owning-ref, until https://github.com/Kimundi/owning-ref-rs/issues/41 is solved.
* The current design first *tags* the text with a markup format, and only *parses* it when given to the view. I think parsing it from the beginning is a better solution, with better error management, so some changes are coming.

So it's still very alpha.

**Comment by gyscos** [comment]
Latest commit fixed a few issues:
* italic text now works on ncurses
* text is first parsed into a `StyledString` (for example using `markdown::parse()`), and *then* passed to the view, where nothing can fail.
* no need for the forked owning-ref

`StyledString` can also be created manually, and concatenated using `StyledString::append`.

**Comment by aleksander** [comment]
Looks strange with termion, crossterm backends:
![screeshot](https://user-images.githubusercontent.com/129811/62004112-98fcf180-b129-11e9-82a1-63dded8038e5.png)

**Comment by gyscos** [comment]
Weird indeed!
Does this also happen with the other examples? It might be caused by a font misalignment... :S

**Comment by aleksander** [comment]
> Weird indeed!
> Does this also happen with the other examples? It might be caused by a font misalignment... :S

No, other examples looks normal.
